### PR TITLE
Implement profile properties

### PR DESCRIPTION
**Pull request by kashike**
[API](https://github.com/SpongePowered/SpongeAPI/pull/974) | [**Common**](https://github.com/SpongePowered/SpongeCommon/pull/344) | [Forge](https://github.com/SpongePowered/SpongeForge/pull/455) | [Vanilla](https://github.com/SpongePowered/SpongeVanilla/pull/194)

### CODE REVIEW

File: src/main/java/org/spongepowered/common/service/user/SpongeUserStorageService.java [hunk 1]
```diff
@@ -55,13 +55,13 @@
...
     public Optional<User> get(GameProfile profile) {
-        return Optional.ofNullable(UserDiscoverer.findByUuid(checkNotNull(checkNotNull(profile, "profile").getUniqueId(), "profile UUID")));
+        return Optional.ofNullable(UserDiscoverer.findByUuid(checkNotNull(checkNotNull(profile, "profile").getUniqueId(), "profile id")));
     }
 
     @Override
     public User getOrCreate(GameProfile profile) {
         if (profile.getUniqueId() == null) {
-            String name = profile.getName();
+            String name = profile.getName().get();
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/service/user/SpongeUserStorageService.java]: `orElse(null)`

File: src/main/java/org/spongepowered/common/mixin/core/ban/MixinUserListBans.java [hunk 2]
```diff
@@ -60,7 +60,7 @@ public UserListEntry getEntry(Object object) {
     public String[] getKeys() {
         List<String> names = new ArrayList<>();
         for (Ban.Profile ban: Sponge.getServiceManager().provideUnchecked(BanService.class).getProfileBans()) {
-            names.add(ban.getProfile().getName());
+            names.add(ban.getProfile().getName().get());
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/mixin/core/ban/MixinUserListBans.java]: should check `isPresent()` before adding

File: src/main/java/org/spongepowered/common/mixin/core/ban/MixinUserListBans.java [hunk 3]
```diff
@@ -83,7 +83,7 @@ public void removeEntry(Object object) {
     @Overwrite
     public com.mojang.authlib.GameProfile isUsernameBanned(String username) {
         for (Ban.Profile ban: Sponge.getServiceManager().provideUnchecked(BanService.class).getProfileBans()) {
-            if (ban.getProfile().getName().equals(username)) {
+            if (ban.getProfile().getName().get().equals(username)) {
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/mixin/core/ban/MixinUserListBans.java]: Check `isPresent`

File: src/main/java/org/spongepowered/common/service/user/SpongeUserStorageService.java (hunk 1 above)
Comment (kashike) [inline src/main/java/org/spongepowered/common/service/user/SpongeUserStorageService.java]: done

File: src/main/java/org/spongepowered/common/mixin/core/ban/MixinUserListBans.java (hunk 2 above)
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/ban/MixinUserListBans.java]: done

File: src/main/java/org/spongepowered/common/mixin/core/ban/MixinUserListBans.java (hunk 3 above)
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/ban/MixinUserListBans.java]: done

File: src/main/java/org/spongepowered/common/data/manipulator/mutable/SpongeRepresentedPlayerData.java [hunk 4]
```diff
@@ -61,10 +64,22 @@ public int getContentVersion() {
...
+
+        @Override
+        public Optional<String> getName() {
+            return Optional.empty();
+        }
+
         @Override
-        public String getName() {
+        public Multimap<String, ProfileProperty> getPropertyMap() {
             return null;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/mutable/SpongeRepresentedPlayerData.java]: Could this return an empty `Multimap`?

File: src/main/java/org/spongepowered/common/mixin/core/authlib/MixinProperty.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(this.getName(), this.getValue(), this.property$getSignature());
+    }
+
+    @Override
+    public String toString() {
+        return Objects.toStringHelper(this)
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/authlib/MixinProperty.java]: Unless you want this to print out as `Property`, I'd recommend actually using the API's defined name as `"ProfileProperty"`

**Comment by gabizou** [comment]
Merged into bleeding in https://github.com/SpongePowered/SpongeCommon/commit/a8740cb622fe0feb9515d98e981d140a321a733f

File: src/main/java/org/spongepowered/common/mixin/core/authlib/MixinProperty.java (hunk 5 above)
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/authlib/MixinProperty.java]: Knowing the actual name is nice

**Pull request closed without merging**